### PR TITLE
Changes for esp-open-sdk

### PR DESCRIPTION
**Pull request by someburner**
## Overview

This is still a bit of a WIP but wanted to get it up for discussion. I made sure it still compiles with the Arduino builder makefile, but should be tested by someone more familiar with Arduino to make sure.

This compiles with current `esp-open-sdk` master. I haven't made a PR there yet- waiting until this guy is cleaned up. But to test you can use my esp-open-sdk fork:

```
git clone --recursive https://github.com/someburner/esp-open-sdk.git -b sdk2-lwip212-new
cd esp-open-sdk
make
```

Or you can do:

```
git clone --recursive https://github.com/pfalcon/esp-open-sdk.git
cd esp-open-sdk
git submodule add https://github.com/someburner/esp82xx-nonos-linklayer-opensdk.git lwip2
cd lwip2
git submodule update --init
```

And then add in the lines to `Makefile` as shown [here](https://github.com/pfalcon/esp-open-sdk/pull/271/files).

## TODO

1. My patches to the SDK includes (user_interface.h, sntp.h) and overwriting ip_addr.h + ipv4_addr.h are a bit ugly, maybe someone has a better idea for this. But they do work for different SDKs, at least on 2.x.

2. Also for `esp-missing.h` there is the issue of the `pvPortxxx` definitions changing between SDKs and being different for Arduino. Are those even necessary to have? Would make life easier.

3. IMO `gluedebug.h` is really messy and hard to follow. Very likely things will break if some of those debug switches are flipped on.

4. Probably most importantly is deciding what to do about SNTP. Currently it expects SNTP to be provided when you compile your project (eg [esp-sntp.c](https://github.com/d-a-v/esp82xx-nonos-linklayer/blob/2101a60213055ff9b689889a8267425bc929764c/glue-lwip/esp-sntp.c)). Personally I like to have control of sntp source in my project, but I could see others wanting that to be compiled in.

### CODE REVIEW

**Comment by someburner** [comment]
## Usage / Gotchas

For me, the resulting SDK is almost a drop-in replacement for esp-open-sdk, but there were a couple things I had to change, so here is a list in case it helps anyone wanting to try this:

1. Remove old sources from project. In my case I had the lwip sources from the SDK `third-party` folder in my project since I had made edits to it. Obviously these will need to be removed.

2. Add esp-sntp.c (or similar) as mentioned above somewhere in your project.

3. Add some definitions to CFLAGS and also link the correct lwip2.a, as there are several that are generated. Here's how mine looks:

```make
################# lwip2 settings #################
# TCP_MSS ?= 536
TCP_MSS ?= 1460

# LWIP_FEAT ?= 0
LWIP_FEAT ?= 1

LWIP_IPV6 ?= 0
# LWIP_IPV6 ?= 1
##################################################

#  Collect lwip2 cflags
LWIP2_CFLAGS := \
	-DTCP_MSS=$(TCP_MSS) \
	-DLWIP_FEATURES=$(LWIP_FEAT) \
	-DLWIP_IPV6=$(LWIP_IPV6)

# Determine which lwip2 lib to link
LWIP2_LIB ?= lwip
ifeq ($(LWIP_IPV6),1)
LWIP2_LIB := $(LWIP2_LIB)6
else
LWIP2_LIB := $(LWIP2_LIB)2
endif
LWIP2_LIB := $(LWIP2_LIB)-$(TCP_MSS)
ifeq ($(LWIP_FEAT),1)
LWIP2_LIB := $(LWIP2_LIB)-feat
endif

# add flags to CCFLAGS (or whatever you use to collect flags)
CCFLAGS += $(LWIP2_CFLAGS)

# where $(OPENSDK_ROOT) is /path/to/esp-open-sdk
CCFLAGS += -I$(OPENSDK_ROOT)/xtensa-lx106-elf/xtensa-lx106-elf/sysroot/usr/include/lwip2

# lastly, wherever your list of libraries to link is (eg -lmain, -lhal, etc)
# add this to list:
-l$(LWIP2_LIB)
```

And then also add these to your `.ld` file:

```
  .irom0.text : ALIGN(4)
  {
    ...

    *liblwip2-536.a:(.literal .text .literal.* .text.*)
    *liblwip2-1460.a:(.literal .text .literal.* .text.*)
    *liblwip2-536-feat.a:(.literal .text .literal.* .text.*)
    *liblwip2-1460-feat.a:(.literal .text .literal.* .text.*)
    *liblwip6-536-feat.a:(.literal .text .literal.* .text.*)
    *liblwip6-1460-feat.a:(.literal .text .literal.* .text.*)

    ...
  } >irom0_0_seg :irom0_0_phdr =0xffffffff
```

**Comment by d-a-v** [comment]
Do you plan to add your last comment in esp-open-sdk's pull request ?
Or is it wise to put it locally in README-esp-open-sdk.txt ?
(I should add arduino's one in README-arduino.txt too)

You can also update the main README.

**Comment by someburner** [comment]
Yes, I'll add those in, thank you.

For SNTP, I was thinking, what if we add in the esp-sntp.c, and make that a makefile switch, same as `esp-millis.c`? We can have it default to enabled for esp-open-sdk so that sntp is usable without needing to include the source themselves. But if they wish, they can compile without the included SNTP and handle the implementation themselves.

**Comment by d-a-v** [comment]
> what if we add in the esp-sntp.c, and make that a makefile switch

I moved it in arduino because we decided that this code here was not coherent.
But I agree, sntp is included in espressif's lwIP-1.4, so it is needed if lwip2 has to be a drop-in replacement.

After that (sntp, espconn, mss=1460, no ipv6), one can try and link this new lwip2 with espressif's libat.a / AT firmware :)

**Comment by someburner** [comment]
Regarding the change for SoftAP, I got that directly from Espressif: https://github.com/espressif/ESP8266_NONOS_SDK/commit/3ba05d37e01f559e97d86de1c45c07f0be43e3b1#diff-a40d96be5c50861565d93be457afd044

From what I can tell they are referencing the same memory location, but shouldn't it be left as `&client_address.addr` for correctness? I believe the struct is just a wrapper around addr like this:

```c
struct ip4_addr {
  u32_t addr;
};
```

**Comment by d-a-v** [comment]
There is no declaration for  `wifi_softap_set_station_info` in nonos-sdk (not in sources, not in api reference).
They do this (from `grep`):
```
struct ip_addr ip_zero;
...
wifi_softap_set_station_info(dhcp_node->mac, &ip_zero);
```
and this conflicts with their
```
wifi_softap_set_station_info(pmsg_dhcps->chaddr, &client_address.addr);
```

The one I added in `esp-missing.h` uses `ipv4_addr*` as second parameter (because `ip_addr` changes with IPv6). I did this for correctness.

It is true we could use `u32*` for the second argument (and `&ip_zero.addr`). I dont mind if you prefer keeping `u32*` but then you must include the signature change from `esp-missing.h`.

**Comment by someburner** [comment]
I see. Yeah it looks like they actually do it both ways now. I'm guessing their signature is u32* but since either works I agree it makes sense to leave as ipv4_addr* to make it clear what kind of address that should be.

**Comment by someburner** [comment]
Okay I added those. Should I go ahead and add the option for SNTP? Maybe that should be added separately.

**Comment by maverickchongo** [comment]
@someburner I am eager to try this in my project, and I wanted to ask you whether you have seen any gain of ram memory af including the LWIP2?

**Comment by maverickchongo** [comment]
@someburner I have tried to build this in the esp-open-sdk (I am using NON-OS-SDK v3) as explained above but get lots of errors regarding pvPortMalloc, like:

`glue-esp/lwip-esp.c:219:75: error: too many arguments to function 'pvPortMalloc'
   struct pbuf_wrapper* p = (struct pbuf_wrapper*)os_malloc((sizeof(struct pbuf_wrapper) * PBUF_WRAPPER_BLOCK));
`

Any idea how to fix this? I guess it is related to a different version of the SDK right?

**Comment by someburner** [comment]
@maverickchongo 

I haven't compared RAM specifically but I think it's roughly the same amount of ram used.

I'm pretty sure I tried compiling against sdk 3.x and got it to work before, but it may not work as-is without some tweaks. The issue may be that in sdk 3.x they changed the `pvPortMalloc`/`zalloc`/etc defs again. You just need to change the definitions in `glue/esp-missing.h` to match the SDK definitions in the NONOS `mem.h` file. And maybe also change the patched file in `lwip2-src/src/core/mem.c` under the `#if MEM_LIBC_MALLOC` definition.

**Comment by maverickchongo** [comment]
Thanks @someburner I'll try that

**Comment by maverickchongo** [comment]
@someburner 

I have added sntp to my source files as suggested but I am getting all sorts of compilation issues, redefinitions of functions etc.

I see some sntp source files in the lwip2 folder, did you finally add them?

**Comment by someburner** [comment]
@maverickchongo - sorry for not getting back. I'm in the process of moving so things are a bit hectic, sometime in a week or 2 I will try building for sdk3.x and get back to you.

**Comment by maverickchongo** [comment]
Thanks @someburner, I really appreciate it.
